Rails tests

stas opened the pull request:
Hi there!

I'm trying to make this work with the latest rails without any luck.
I added some tests to replicate the issue, but didn't take time yet to look into what's wrong.
So, for now, I'll just share this until I/we figure what's the fix (feel free to share any details I might be missing).

Aside from this, I'd like to suggest to remove `dry-configurable` from dependencies, since we really do not need that.

I also tweaked the way the executable is discovered to make it work on systems without `bash` (ex. Alpine).

Thanks for all the work so far!

stas committed:
Use plain ruby to handle configuration.

stas committed:
Do not rely on bash. It's not available always.

stas committed:
Added a rails test.

m-basov commented:
Hey @stas! I like the changes and will try to find some time for review and test the issue in the nearest 2 weeks(most likely 7-8 Nov).

> I also tweaked the way the executable is discovered to make it work on systems without bash (ex. Alpine).

That's cool! I created this gem a long time ago when I didn't know that bash is not POSIX compliant 😄 

> Aside from this, I'd like to suggest to remove dry-configurable from dependencies, since we really do not need that.

I love reduced dependencies list as well!

stas committed:
No more post install messages.

Aside from this being super annoying, it's really not helpful...

stas committed:
When in Rails, use Rails...

stas commented:
Ok, I figured out what was wrong.

Also locally, it's really important to have the `gem 'mjml-ruby', require: 'mjml'` otherwise, things just do not work :shrug: 

I also removed the post install message since it's super annoying and loses relevancy over time (not to mention that we already raise an exception, so it's absolutely extra).

As well, on rails, I moved the `MJML.setup!` into the intializer to keep things rails.


Long term, I was looking at the way [HAML does it](https://github.com/haml/haml-rails/blob/master/lib/haml-rails.rb), and I think we should set up a caching mechanism, where templates are being cached to speed up and avoid going back and forth between nodejs and rails. I'll try to take a look at it separately though.

stas committed:
Carry the error message into the exception.

stas commented:
The error messages should also carry on now better.

stas committed:
Too much logging.

This polutes the development logs and generates massive log files in
test mode on Rails. The gem already provides enough API to debug and
solve invalid templates, so all of this is very much extra.

m-basov commented on mjml-ruby.gemspec:
```diff
@@ -11,16 +11,16 @@ Gem::Specification.new do |s|
-  s.add_runtime_dependency 'dry-configurable', '~> 0.1', '>= 0.1.3'
-
   s.add_development_dependency 'rake'
+  s.add_development_dependency 'railties'
+  s.add_development_dependency 'actionmailer'
+  s.add_development_dependency 'tzinfo-data'
```
Could you add please versions constraints?

stas replied:
@m-basov sure, up to you. I generally leave the dev deps unconstrained to avoid the noise or issues/prs to "bump versions". And since these are dev deps, the users don't care.

m-basov replied:
I usually specify at least major version to avoid random breakage on CI.

m-basov commented on lib/mjml/parser.rb:
```diff
@@ -29,20 +26,21 @@ def call!(template)
       out, err = should_get_outpout_from_file? ? output_from_file : output_from_memory
       parsed = parse_output(out)
 
-      MJML.logger.debug("Output:\n #{parsed[:output]}")
```
Not sure about logging. Could you explain please while it is not helpful? As consumer you always can silence them by changing logger level. As maintainer it is helpful to have them during the debug process.

stas replied:
Actually a couple of reasons:
 * the main one is that mjml works more like a transpiler rather than a generator, the code it generates is arguably readable and it ends up either way in your rendered template (either mailer preview or the page that was rendered, which ultimately are superior to inspect). In a nutshell, if haml doesn't do it, why should we do it?!
 * if the templates are parsed properly, nobody cares what's the output, if there's an error, the changes I added related to the error messages clearly state where the issues are.
 * there's no way to disable this in dev/test environment, and you end up with gigabytes of log data, not to mention it that it makes important logs messages harder to find (a template generates hundreds of lines of output in the log, to the point that the main request log message is lost and requires extra effort to scroll around and find it).

Long story short, we either separate the debugging from the main logger, or we remove it because it harms more than it helps now.

Let me know what makes you want to have it like it is now?

m-basov replied:
Interesting. Default Logger level is error so debug messages shouldn't appear at all unless you specify your custom logger with another logger level. https://github.com/m-basov/mjml-ruby/blob/205aa2f9d9f0a58e0a4f390605a2aac824e20fde/lib/mjml/logger.rb#L8

m-basov replied:
I don't have very strong opinion on this so if my comment above doesn't work feel free to merge.

m-basov reviewed (commented):
PR looks great! :heart: 

Could you please rebase it on latest master(I've moved from Travis to Github Actions)? 

I am ready to merge it when my questions will be resolved/answered. 

Also, if you are actively using this gem please let me know I can add you as collaborator since I don't use it already and maintain it only when someone reports an issue or creates a PR.

> Also locally, it's really important to have the `gem 'mjml-ruby', require: 'mjml'` otherwise, things just do not work

Yes, it is mentioned in README. It is because the gem's name is `mjml-ruby` but I name it internally `mjml`. Unfortunately `mjml` name was already taken when I created this gem.

stas commented:
@m-basov would you be kind to update this line into `on: [push, pull_request]`:
https://github.com/m-basov/mjml-ruby/blob/master/.github/workflows/ci.yaml#L3

This will allow the PRs' to report the tests status. It's really weird, but I could not figure out another way to enable tests on forks...

Thanks for reviewing this!

m-basov commented:
@stas Ooops, fixed!

stas commented:
Thanks for adding me to the collaborators @m-basov :bow:

m-basov commented:
Published v0.4.0. Thank you very much for the contribution!
